schema.c

```diff
@@ -5193,3 +5383,242 @@ sch_translate_copy_query (GNode *query)
 
     return copy;
 }
+
+static bool
+_sch_xml_has_proxy (sch_instance *instance, sch_node * schema, xmlNs *ns,
+                    xmlNode * xml, int depth, int flags)
+{
+    char *name = (char *) xml->name;
+    xmlNode *child;
+    bool found = false;
+
+    /* Detect change in namespace */
+    if (xml->ns && xml->ns->href)
+    {
+         xmlNs *nns = sch_lookup_ns (instance, schema, (const char *) xml->ns->href, flags, true);
+         if (nns)
+            ns = nns;
+    }
+
+    /* Find schema node */
+    if (sch_is_proxy (schema))
+        return true;
+
+    if (!schema)
+    {
+        schema = xmlDocGetRootElement (instance->doc);
+        /* Detect change in namespace with the new schema */
+        if (xml->ns && xml->ns->href)
+        {
+             xmlNs *nns = sch_lookup_ns (instance, schema, (const char *) xml->ns->href, flags, true);
+             if (nns)
+                ns = nns;
+        }
+    }
+    schema = _sch_node_child (ns, schema, name);
+    if (schema == NULL)
+        return false;
+
+    /* Prepend non default namespaces to root nodes */
+    if (depth == 0 && ns && ns->prefix && !sch_ns_native (instance, ns))
+        name = g_strdup_printf ("%s:%s", ns->prefix, (const char *) xml->name);
+    else
+        name = g_strdup ((char *) xml->name);
+
+    /* LIST */
+    if (sch_is_list (schema))
+    {
+        DEBUG (flags, "%*s%s%s\n", depth * 2, " ", depth ? "" : "/", name);
+        depth++;
+        schema = sch_node_child_first (schema);
+    }
+
+    for (child = xmlFirstElementChild (xml); child; child = xmlNextElementSibling (child))
+        found = _sch_xml_has_proxy (instance, schema, ns, child, depth + 1, flags);
```

Comment:
I can't make much sense of a lot of this, but this looks a bit suspicious to me. Maybe should be OR of all found values or AND of them all?

Reply:
I have modified the for loop to exit as soon as found is set true. This will cascade back up to the top of the recursive routines. The proxy detection routines are based on the equivalent _sch_xml_to_gnode and _sch_path_to_gnode routines with the setting of the GNode tree nodes removed, so its not actually new code.